cannon/mipsevm/tests/difftester.go

```diff
@@ -0,0 +1,201 @@
+package tests
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/arch"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded"
+	mtutil "github.com/ethereum-optimism/optimism/cannon/mipsevm/multithreaded/testutil"
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/testutil"
+)
+
+type TestCase interface {
+	Name() string
+}
+
+type InitializeStateFn[T TestCase] func(testCase T, state *multithreaded.State, vm VersionedVMTestCase)
+type SetExpectationsFn[T TestCase] func(testCase T, expect *mtutil.ExpectedState)
+
+type DiffTester[T TestCase] struct {
+	stateOpts       []mtutil.StateOption
+	initState       InitializeStateFn[T]
+	setExpectations SetExpectationsFn[T]
+}
+
+func NewDiffTester[T TestCase]() *DiffTester[T] {
+	return &DiffTester[T]{}
+}
+
+func (d *DiffTester[T]) InitState(initStateFn InitializeStateFn[T], opts ...mtutil.StateOption) *DiffTester[T] {
+	d.initState = initStateFn
+	d.stateOpts = opts
+
+	return d
+}
+
+func (d *DiffTester[T]) SetExpectations(setExpectationsFn SetExpectationsFn[T]) *DiffTester[T] {
+	d.setExpectations = setExpectationsFn
+
+	return d
+}
+
+func (d *DiffTester[T]) Run(t *testing.T, testCases []T, randSeed int64, opts ...TestOption) {
```

Comment:
nit: The Run function could be simplified for callers. All inputs except maybe `t`, so that a `DiffTester` can be reused, could be preconfigured. Similar to how `setExpectations` is configured.
The seed could the have a default value if it's not configured.

Reply:
I think the current `Run` signature holds all of the values I wouldn't expect to be shared across test runs, so that we could reuse the DiffTester (which just remembers initialization / expectation logic).  And I kind of like forcing the caller to set a random seed so we get some variety across tests even though it's a little verbose.  

But happy to iterate / discuss more.